src/services/suggestionDiagnostics.ts

```diff
@@ -109,7 +113,64 @@ namespace ts {
         }
     }
 
+    function addConvertToAsyncFunctionDiagnostics(node: FunctionLikeDeclaration, checker: TypeChecker, diags: DiagnosticWithLocation[]): void {
+
+        const functionType = node.type ? checker.getTypeFromTypeNode(node.type) : undefined; 
+        if (isAsyncFunction(node) || !node.body || !functionType) {
+            return;
+        }
+
+        const callSignatures = checker.getSignaturesOfType(functionType, SignatureKind.Call); 
+        const returnType = callSignatures.length ? checker.getReturnTypeOfSignature(callSignatures[0]) : undefined;
+
+        if (!returnType || !checker.getPromisedTypeOfPromise(returnType)) {
+            return;
+        }
+
+        // collect all the return statements
+        // check that a property access expression exists in there and that it is a handler
+        const returnStatements = getReturnStatementsWithPromiseCallbacks(node);
+        if (returnStatements.length > 0) {
+            diags.push(createDiagnosticForNode(isVariableDeclaration(node.parent) ? node.parent.name : node, Diagnostics.This_may_be_converted_to_an_async_function));

```

Comment:
`node` seems like a pretty big span.  You might want to check where we report other function suggestions - I think it's just on the name or signature.